### PR TITLE
Implement enhanced API usability improvements for issue #857

### LINKED ISSUE #857: Make iCalendar easier to use and more efficient
iCalendar currently requires a lot of digging to create a valid calenda:

- it requires the user to know which properties to add to which components, e.g. `dtstart` for the start time of an event
- it provides neither guidance nor enforcement as to which properties are required

When used for creating calendars, the value of iCalendar compared to generating the raw byte string is some, but not huge.

```python
from icalendar import Calendar, Event
import zoneinfo
from datetime import datetime

# Developer must know both of these fields are required, and what's their required format and value
cal = Calendar()
cal.add('prodid', '-//My calendar product//mxm.dk//')
cal.add('version', '2.0')

event = Event()
# Developer must know all of these fields are required
event.add('summary', 'Python meeting about calendaring')
event.add('dtstart', datetime(2005,4,4,8,0,0,tzinfo=zoneinfo.ZoneInfo("UTC")))
event.add('dtend', datetime(2005,4,4,10,0,0,tzinfo=zoneinfo.ZoneInfo("UTC")))
event.add('dtstamp', datetime(2005,4,4,0,10,0,tzinfo=zoneinfo.ZoneInfo("UTC")))

cal.add_component(event)
```

For all of the code above, here's the raw iCal output:

```
BEGIN:VCALENDAR
VERSION:2.0
PRODID:-//My calendar product//mxm.dk//
BEGIN:VEVENT
SUMMARY:Python meeting about calendaring
DTSTART:20050404T080000Z
DTEND:20050404T100000Z
DTSTAMP:20050404T001000Z
END:VEVENT
END:VCALENDAR
```

iCalendar's code actually exceeds the raw iCalendar output in amount and complexity.

## Expected behavior

When used in generative mode, iCalendar should vastly simplify generation:

- internally set all required fields with known values (e.g. "version")
- require all fields that are required but lack a meaningful defaults in the object's constructor
- expose intuitive names as attributes of the objects
- internally ensure that the output is valid, which should be granted enabled by the provisions above. For example, `to_ics()` should raise an exception on a calendar lacking an event

Here's an example:

```python
from icalendar import Calendar, Event
import zoneinfo
from datetime import datetime, timedelta

# automatically generates prodid='-//foo.org//My parties//EN' and uid = a sha256 of it
cal = Calendar(organization="foo.org", name="My parties", language="en")

# automatically generates dtstamp = now() and uid = sha256 of str(now())
event = Event(start=mydatetime)
event.end = mydatetime + timedelta(minutes=15)
event.name = "Some string"
```

I am no expert in RFC 5545 and the examples above might be inaccurate. I am merely suggesting to smarten up the interface of the icalendar package so it conceals the raw syntax of iCalendar and allows the user to operate in a more intuitive space -- at least for the fundamental items:

- cover Calendar + Events. All other components are more sophisticated
- title of the event (summary)
- start time
- duration
- location
- URL

## Related issues

Several other issues overlap with this:

- https://github.com/collective/icalendar/issues/843
- https://github.com/collective/icalendar/issues/841
- https://github.com/collective/icalendar/issues/740

### PR DESCRIPTION
**Pull request by SashankBhamidi**
Fixes #857 by implementing the remaining API usability enhancements requested in the issue.

## Problem
Issue #857 identified several pain points with the current iCalendar API:
- Users must know specific RFC field names (PRODID, VERSION, DTSTART, etc.)
- Manual calculation of end times from start + duration
- No validation to ensure calendars are complete before output
- Complex workflow requiring extensive RFC knowledge

## Solution
Building on the foundation from merged PR #845, this implements the remaining enhancements:

### 1. Organization-based PRODID Generation
```python
# Before: Manual PRODID specification
cal = Calendar()
cal.add('prodid', '-//My Org//My App//EN')

# After: Automatic generation from organization/language
cal = Calendar.new(organization="example.org", name="My App", language="en")
# Generates: PRODID:-//example.org//My App//EN
```

### 2. Intuitive Duration Setting
```python
# Before: Manual end time calculation
event = Event.new(start=datetime(2026, 3, 19, 12, 30))
event.end = datetime(2026, 3, 19, 14, 30)  # User calculates

# After: Automatic end time from duration
event = Event.new(start=datetime(2026, 3, 19, 12, 30))
event.duration = timedelta(hours=2)  # Automatically sets end time
```

### 3. Explicit Validation
```python
# New validation method
calendar.validate()  # Raises IncompleteComponent if invalid
```

## User's Desired Workflow (Now Working\!)
The exact API requested in issue #857:
```python
cal = Calendar.new(organization="foo.org", name="My parties", language="en")
event = Event.new(start=mydatetime)
event.duration = timedelta(minutes=15)
event.summary = "Some string"
cal.add_component(event)
```

## Changes
- **Modified**: `src/icalendar/cal/calendar.py` - Enhanced Calendar.new() with organization/language support + validate() method
- **Modified**: `src/icalendar/cal/event.py` - Added duration setter for automatic end time calculation  
- **Added**: `src/icalendar/tests/test_issue_857_enhanced_api.py` - Comprehensive test coverage (14 tests)
- **Updated**: `CHANGES.rst` - Documented all enhancements

## Backward Compatibility
- ✅ All existing functionality preserved
- ✅ 453 existing tests still pass
- ✅ No breaking changes

## Test Coverage
- Organization-based PRODID generation scenarios
- Duration setter with various timezones and edge cases
- Validation method behavior
- Complete user workflow from issue #857
- Cross-platform compatibility (Python 3.8-3.13, nopytz environments)

----
📚 Documentation preview 📚: https://icalendar--879.org.readthedocs.build/

### CODE REVIEW

**Comment by SashankBhamidi** [comment]
@mij - This PR implements the exact API you requested in #857. Your examples now work as specified.

@niccokunzmann @tobixen - Thanks for the guidance in #857 and the foundation from #845.

File: CHANGES.rst [hunk 1]
```diff
@@ -19,6 +19,9 @@ Minor changes:
 - Add type annotation to ``from_ical()``.
 - Fix enum documentation.
 - ``DTSTAMP``, ``LAST_MODIFIED``, and ``CREATED`` can now be set to ``None`` to delete the value.
+- Enhanced ``Calendar.new()`` to support organization and language parameters for automatic PRODID generation.
```
Comment (stevepiercy) [inline CHANGES.rst]:
```suggestion
- Enhanced ``Calendar.new()`` to support organization and language parameters for automatic ``PRODID`` generation.
```

File: src/icalendar/cal/calendar.py [hunk 2]
```diff
@@ -457,24 +458,29 @@ def new(
...
 
-        This creates a new Todo in accordance with :rfc:`5545`.
+        This creates a new Calendar in accordance with :rfc:`5545`.
 
         Arguments:
             calscale: The :attr:`calscale` of the component.
             categories: The :attr:`categories` of the component.
             color: The :attr:`color` of the component.
             description: The :attr:`description` of the component.
+            language: The language for the calendar. Used to generate localized prodid.
```
Comment (stevepiercy) [inline src/icalendar/cal/calendar.py]:
```suggestion
            language: The language for the calendar. Used to generate localized ``prodid``.
```

File: src/icalendar/cal/calendar.py [hunk 3]
```diff
@@ -457,24 +458,29 @@ def new(
...
             method: The :attr:`method` of the component.
             name: The :attr:`calendar_name` of the component.
-            prodid: The :attr:`prodid` of the component.
+            organization: The organization name. Used to generate prodid if not provided.
```
Comment (stevepiercy) [inline src/icalendar/cal/calendar.py]:
```suggestion
            organization: The organization name. Used to generate ``prodid`` if not provided.
```

File: src/icalendar/cal/calendar.py [hunk 4]
```diff
@@ -457,24 +458,29 @@ def new(
...
+            prodid: The :attr:`prodid` of the component. If None and organization is provided,
+                generates a prodid in format "-//organization//name//language".
```
Comment (stevepiercy) [inline src/icalendar/cal/calendar.py]:
```suggestion
            prodid: The :attr:`prodid` of the component. If ``None`` and organization is provided,
                generates a ``prodid`` in the format of ``-//organization//name//language``.
```

File: src/icalendar/cal/event.py [hunk 5]
```diff
@@ -307,10 +307,23 @@ def duration(self) -> timedelta:
...
+        When setting duration, the end time is automatically calculated from start + duration.
         """
         return self.end - self.start
 
+    @duration.setter
+    def duration(self, duration: timedelta):
+        """Set the duration of the event.
+        
+        This automatically calculates and sets the end time as start + duration.
+        If no start time is set, raises IncompleteComponent.
```
Comment (stevepiercy) [inline src/icalendar/cal/event.py:319]:
```suggestion
        If no start time is set, raises ``IncompleteComponent``.
```

File: src/icalendar/tests/test_issue_857_enhanced_api.py [hunk 6]
```diff
@@ -0,0 +1,214 @@
...
+        name="My parties",
+        language="en"
+    )
+    
+    assert calendar.calendar_name == "My parties"
+    assert calendar.prodid == "-//foo.org//My parties//EN"
+
+
+def test_calendar_new_with_organization_no_language():
+    """Test Calendar.new() with organization but no language defaults to EN."""
```
Comment (stevepiercy) [inline src/icalendar/tests/test_issue_857_enhanced_api.py]:
```suggestion
    """Test Calendar.new() with organization but no language. Defaults to EN."""
```

File: src/icalendar/tests/test_issue_857_enhanced_api.py [hunk 7]
```diff
@@ -0,0 +1,214 @@
...
+    calendar = Calendar.new(
+        organization="example.com",
+        name="Test Calendar"
+    )
+    
+    assert calendar.prodid == "-//example.com//Test Calendar//EN"
+
+
+def test_calendar_new_with_organization_no_name():
+    """Test Calendar.new() with organization but no name uses 'Calendar'."""
```
Comment (stevepiercy) [inline src/icalendar/tests/test_issue_857_enhanced_api.py]:
```suggestion
    """Test Calendar.new() with organization but no name. Uses 'Calendar'."""
```

**Review by stevepiercy — CHANGES_REQUESTED** [review]
Minor markup tweaks. Needs technical review. Also suggest running `tox -e ruff` on Python files for code quality.

**Comment by stevepiercy** [comment]
> Also suggest running `tox -e ruff` on Python files for code quality.

@SashankBhamidi please ignore that bad suggestion from me. We need to better document how to use it for contributions first. We only need proper formatting on the files you touch, not the entire code base.

**Comment by SashankBhamidi** [comment]
> > Also suggest running `tox -e ruff` on Python files for code quality.
> 
> @SashankBhamidi please ignore that bad suggestion from me. We need to better document how to use it for contributions first. We only need proper formatting on the files you touch, not the entire code base.

I've updated all PRs with ruff formatting applied only to the feature-specific files as eequested. Ready for technical review.

File: src/icalendar/cal/calendar.py [hunk 8]
```diff
@@ -499,5 +514,22 @@ def new(
         calendar.uid = uid
         return calendar
 
+    def validate(self):
```
Comment (niccokunzmann) [inline src/icalendar/cal/calendar.py:517]: Nice!

I think, a validate method is actually a good thing to have.

File: src/icalendar/cal/calendar.py [hunk 9]
```diff
@@ -488,6 +494,15 @@ def new(
         .. warning:: As time progresses, we will be stricter with the validation.
         """
         calendar = cls()
+
```
Comment (niccokunzmann) [inline src/icalendar/cal/calendar.py:497]: Nice!

File: src/icalendar/cal/event.py [hunk 10]
```diff
@@ -307,10 +307,28 @@ def duration(self) -> timedelta:
...
+        if (
+            not hasattr(self, "_get_start_end_duration")
+            or self._get_start_end_duration()[0] is None
+        ):
+            raise IncompleteComponent(
+                "Cannot set duration without DTSTART. Set start time first."
+            )
+
+        start_time = self.start
+        self.end = start_time + duration
```
Comment (niccokunzmann) [inline src/icalendar/cal/event.py:330]: As a note: this does not always work with pytz. It works well for events.

File: src/icalendar/tests/test_issue_857_enhanced_api.py [hunk 11]
```diff
@@ -0,0 +1,202 @@
...
+
+
+def test_calendar_validate_no_components():
+    """Test that validate() detects empty calendar."""
+    calendar = Calendar()
+    calendar.add("PRODID", "-//Test//Test//EN")
+    calendar.add("VERSION", "2.0")
+
+    with pytest.raises(
+        IncompleteComponent, match="Calendar must contain at least one component"
```
Comment (niccokunzmann) [inline src/icalendar/tests/test_issue_857_enhanced_api.py:118]: Yay!

**Review by niccokunzmann — APPROVED** [review]
Thank you very much! This looks good!
